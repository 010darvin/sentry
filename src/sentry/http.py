"""
sentry.utils.http
~~~~~~~~~~~~~~~~~

:copyright: (c) 2010-2014 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
from __future__ import absolute_import

import sentry

import ipaddress
import six
import socket
import requests
import warnings
import time
import logging

from sentry import options
from django.core.exceptions import SuspiciousOperation
from collections import namedtuple
from django.conf import settings
from requests.adapters import HTTPAdapter
from requests.exceptions import SSLError, RequestException, Timeout
from six.moves.urllib.parse import urlparse

from sentry.models import EventError
from sentry.exceptions import RestrictedIPAddress
from sentry.utils.cache import cache
from sentry.utils.hashlib import md5_text
from sentry.utils.strings import truncatechars


logger = logging.getLogger(__name__)

# TODO(dcramer): we want to change these to be constants so they are easier
# to translate/link again
# the maximum number of remote resources (i.e. sourc eifles) that should be
# fetched
MAX_URL_LENGTH = 150

# UrlResult.body **must** be bytes
UrlResult = namedtuple('UrlResult',
    ['url', 'headers', 'body', 'status', 'encoding']
)


# In case SSL is unavailable (light builds) we can't import this here.
try:
    from OpenSSL.SSL import ZeroReturnError
except ImportError:
    class ZeroReturnError(Exception):
        pass

USER_AGENT = 'sentry/{version} (https://sentry.io)'.format(
    version=sentry.VERSION,
)

DISALLOWED_IPS = {
    ipaddress.ip_network(six.text_type(i), strict=False)
    for i in settings.SENTRY_DISALLOWED_IPS
}


class BadSource(Exception):
    error_type = EventError.UNKNOWN_ERROR

    def __init__(self, data=None):
        if data is None:
            data = {}
        data.setdefault('type', self.error_type)
        super(BadSource, self).__init__(data['type'])
        self.data = data


class CannotFetch(BadSource):
    error_type = EventError.FETCH_GENERIC_ERROR


def get_server_hostname():
    return urlparse(options.get('system.url-prefix')).hostname


def is_valid_url(url):
    """
    Tests a URL to ensure it doesn't appear to be a blacklisted IP range.
    """
    # If we have no disallowed ips, we can skip any further validation
    # and there's no point in doing a DNS lookup to validate against
    # an empty list.
    if not DISALLOWED_IPS:
        return True

    parsed = urlparse(url)
    if not parsed.hostname:
        return False

    server_hostname = get_server_hostname()

    if parsed.hostname == server_hostname:
        return True

    # NOTE: The use of `socket.gethostbyname` is slightly flawed.
    # `gethostbyname` doesn't handle octal IP addresses correctly, nor
    # does it fetch all of the IP addresses for the record.
    # `getaddrinfo` does the correct thing with octals here, and also returns all
    # ip addresses for the hostname.
    #
    # WARNING: This behavior is only correct on Linux. On OSX, `getaddrinfo` also
    # returns the wrong IP.
    #
    # The following should all technically resolve to `127.0.0.1`:
    # Python 2.7.11 Linux
    # >>> socket.gethostbyname('0177.0000.0000.0001')
    # '177.0.0.1'
    # >>> socket.getaddrinfo('0177.0000.0000.0001', 0)[0]
    # (2, 1, 6, '', ('127.0.0.1', 0))
    # Python 2.7.11 macOS
    # >>> socket.gethostbyname('0177.0000.0000.0001')
    # '177.0.0.1'
    # >>> socket.getaddrinfo('0177.0000.0000.0001', None)[0]
    # (2, 2, 17, '', ('177.0.0.1', 0))
    try:
        ip_addresses = set(addr for _, _, _, _, addr in socket.getaddrinfo(parsed.hostname, 0))
    except socket.gaierror:
        return False

    for addr in ip_addresses:
        ip_address = addr[0]
        if ip_address == server_hostname:
            return True

        ip_address = ipaddress.ip_address(six.text_type(ip_address))
        for ip_network in DISALLOWED_IPS:
            if ip_address in ip_network:
                return False

    return True


class BlacklistAdapter(HTTPAdapter):
    def send(self, request, *args, **kwargs):
        if not is_valid_url(request.url):
            raise RestrictedIPAddress('%s matches the URL blacklist' % (request.url,))
        return super(BlacklistAdapter, self).send(request, *args, **kwargs)


def build_session():
    session = requests.Session()
    session.headers.update({'User-Agent': USER_AGENT})
    session.mount('https://', BlacklistAdapter())
    session.mount('http://', BlacklistAdapter())
    return session


def safe_urlopen(url, method=None, params=None, data=None, json=None,
                 headers=None, allow_redirects=False, timeout=30,
                 verify_ssl=True, user_agent=None, session=None):
    """
    A slightly safer version of ``urlib2.urlopen`` which prevents redirection
    and ensures the URL isn't attempting to hit a blacklisted IP range.
    """
    if user_agent is not None:
        warnings.warn('user_agent is no longer used with safe_urlopen')

    if session is None:
        session = build_session()

    kwargs = {}

    if json:
        kwargs['json'] = json
        if not headers:
            headers = {}
        headers.setdefault('Content-Type', 'application/json')

    if data:
        kwargs['data'] = data

    if params:
        kwargs['params'] = params

    if headers:
        kwargs['headers'] = headers

    if method is None:
        method = 'POST' if (data or json) else 'GET'

    try:
        response = session.request(
            method=method,
            url=url,
            allow_redirects=allow_redirects,
            timeout=timeout,
            verify=verify_ssl,
            **kwargs
        )
    # Our version of requests does not transform ZeroReturnError into an
    # appropriately generically catchable exception
    except ZeroReturnError as exc:
        import sys
        exc_tb = sys.exc_info()[2]
        six.reraise(SSLError, exc, exc_tb)
        del exc_tb

    # requests' attempts to use chardet internally when no encoding is found
    # and we want to avoid that slow behavior
    if not response.encoding:
        response.encoding = 'utf-8'

    return response


def safe_urlread(response):
    return response.content


def expose_url(url):
    if url is None:
        return u'<unknown>'
    if url[:5] == 'data:':
        return u'<data url>'
    url = truncatechars(url, MAX_URL_LENGTH)
    if isinstance(url, six.binary_type):
        url = url.decode('utf-8', 'replace')
    return url


def fetch_file(url, headers=None, cache_enabled=True):
    """
    Pull down a URL, returning a UrlResult object.

    Attempts to fetch from the cache.
    """
    # lock down domains that are problematic
    if cache_enabled:
        domain = urlparse(url).netloc
        domain_key = 'source:blacklist:v2:%s' % (
            md5_text(domain).hexdigest(),
        )
        domain_result = cache.get(domain_key)
        if domain_result:
            domain_result['url'] = url
            raise CannotFetch(domain_result)

    logger.debug('Fetching %r from the internet', url)

    http_session = build_session()
    response = None
    try:
        try:
            start = time.time()
            response = http_session.get(
                url,
                allow_redirects=True,
                verify=False,
                headers=headers,
                timeout=settings.SENTRY_FETCH_SOCKET_TIMEOUT,
                stream=True,
            )

            try:
                cl = int(response.headers['content-length'])
            except (LookupError, ValueError):
                cl = 0
            if cl > settings.SENTRY_FETCH_MAX_SIZE:
                raise OverflowError()

            contents = []
            cl = 0

            # Only need to even attempt to read the response body if we
            # got a 200 OK
            if response.status_code == 200:
                for chunk in response.iter_content(16 * 1024):
                    if time.time() - start > settings.SENTRY_FETCH_TIMEOUT:
                        raise Timeout()
                    contents.append(chunk)
                    cl += len(chunk)
                    if cl > settings.SENTRY_FETCH_MAX_SIZE:
                        raise OverflowError()

        except Exception as exc:
            logger.debug('Unable to fetch %r', url, exc_info=True)
            if isinstance(exc, RestrictedIPAddress):
                error = {
                    'type': EventError.RESTRICTED_IP,
                    'url': expose_url(url),
                }
            elif isinstance(exc, SuspiciousOperation):
                error = {
                    'type': EventError.SECURITY_VIOLATION,
                    'url': expose_url(url),
                }
            elif isinstance(exc, Timeout):
                error = {
                    'type': EventError.FETCH_TIMEOUT,
                    'url': expose_url(url),
                    'timeout': settings.SENTRY_FETCH_TIMEOUT,
                }
            elif isinstance(exc, OverflowError):
                error = {
                    'type': EventError.FETCH_TOO_LARGE,
                    'url': expose_url(url),
                    # We want size in megabytes to format nicely
                    'max_size': float(settings.SENTRY_FETCH_MAX_SIZE) / 1024 / 1024,
                }
            elif isinstance(exc, (RequestException, ZeroReturnError)):
                error = {
                    'type': EventError.FETCH_GENERIC_ERROR,
                    'value': six.text_type(type(exc)),
                    'url': expose_url(url),
                }
            else:
                logger.exception(six.text_type(exc))
                error = {
                    'type': EventError.UNKNOWN_ERROR,
                    'url': expose_url(url),
                }

            # TODO(dcramer): we want to be less aggressive on disabling domains
            if cache_enabled:
                cache.set(domain_key, error or '', 300)
            logger.warning('source.disabled', extra=error)
            raise CannotFetch(error)

        body = b''.join(contents)
        headers = {k.lower(): v for k, v in response.headers.items()}
        encoding = response.encoding

        result = (headers, body, response.status_code, encoding)
    finally:
        if response is not None:
            response.close()

    if result[2] != 200:
        logger.debug('HTTP %s when fetching %r', result[2], url,
                     exc_info=True)
        error = {
            'type': EventError.JS_INVALID_HTTP_CODE,
            'value': result[2],
            'url': expose_url(url),
        }
        raise CannotFetch(error)

    # Make sure the file we're getting back is six.binary_type. The only
    # reason it'd not be binary would be from old cached blobs, so
    # for compatibility with current cached files, let's coerce back to
    # binary and say utf8 encoding.
    if not isinstance(result[1], six.binary_type):
        try:
            result = (result[0], result[1].encode('utf8'), None)
        except UnicodeEncodeError:
            error = {
                'type': EventError.JS_INVALID_SOURCE_ENCODING,
                'value': 'utf8',
                'url': expose_url(url),
            }
            raise CannotFetch(error)

    return UrlResult(url, result[0], result[1], response.status_code, result[3])
