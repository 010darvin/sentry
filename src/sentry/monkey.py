from __future__ import absolute_import


def register_scheme(name):
    from six.moves.urllib import parse as urlparse
    uses = urlparse.uses_netloc, urlparse.uses_query, urlparse.uses_relative, urlparse.uses_fragment
    for use in uses:
        if name not in use:
            use.append(name)


register_scheme('app')
