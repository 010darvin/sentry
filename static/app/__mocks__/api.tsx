import isEqual from 'lodash/isEqual';

import type {HandleRequestErrorOptions, RequestOptions, ResponseMeta} from 'sentry/api';
import ModalStore from 'sentry/stores/modalStore';

import {SUDO_REQUIRED, SUPERUSER_REQUIRED} from '../constants/apiErrorCodes';

export class Request {}

export const initApiClientErrorHandling = () => false;
export const hasProjectBeenRenamed = (_args: any[]) => false;

const respond = (isAsync: boolean, fn?: Function, ...args: any[]): void => {
  if (!fn) {
    return;
  }

  if (isAsync) {
    setTimeout(() => fn(...args), 1);
    return;
  }

  fn(...args);
};

type FunctionCallback<Args extends any[] = any[]> = (...args: Args) => void;

/**
 * Callables for matching requests based on arbitrary conditions.
 */
interface MatchCallable {
  (url: string, options: RequestOptions): boolean;
}

type ResponseType = ResponseMeta & {
  body: any;
  callCount: 0;
  headers: Record<string, string>;
  match: MatchCallable[];
  method: string;
  statusCode: number;
  url: string;
};

type MockResponse = [resp: ResponseType, mock: jest.Mock];

/**
 * Compare two records. `want` is all the entries we want to have the same value in `check`
 */
function compareRecord(want: Record<string, any>, check: Record<string, any>): boolean {
  for (const key in want) {
    if (!isEqual(check[key], want[key])) {
      return false;
    }
  }
  return true;
}

afterEach(() => {
  // if any errors are caught we console.error them
  const errors = Object.values(Client.errors);
  if (errors.length > 0) {
    for (const err of errors) {
      // eslint-disable-next-line no-console
      console.error(err);
    }
    Client.errors = {};
  }
});

class Client implements Client {
  static mockResponses: MockResponse[] = [];

  static mockAsync = false;

  static clearMockResponses() {
    Client.mockResponses = [];
  }

  /**
   * Create a query string match callable.
   *
   * Only keys/values defined in `query` are checked.
   */
  static matchQuery(query: Record<string, any>): MatchCallable {
    const queryMatcher: MatchCallable = (_url, options) => {
      return compareRecord(query, options.query ?? {});
    };

    return queryMatcher;
  }

  /**
   * Create a data match callable.
   *
   * Only keys/values defined in `data` are checked.
   */
  static matchData(data: Record<string, any>): MatchCallable {
    const dataMatcher: MatchCallable = (_url, options) => {
      return compareRecord(data, options.data ?? {});
    };

    return dataMatcher;
  }

  // Returns a jest mock that represents Client.request calls
  static addMockResponse(response: Partial<ResponseType>) {
    const mock = jest.fn();

    Client.mockResponses.unshift([
      {
        url: '',
        status: 200,
        statusCode: 200,
        statusText: 'OK',
        responseText: '',
        responseJSON: '',
        body: '',
        method: 'GET',
        callCount: 0,
        match: [],
        ...response,
        headers: response.headers ?? {},
        getResponseHeader: (key: string) => response.headers?.[key] ?? null,
        rawResponse: {
          headers: new Headers(),
          ok: true,
          redirected: false,
          status: 200,
          statusText: 'OK',
          url: 'http://localhost',
          bodyUsed: false,
          body: {
            locked: false,
            cancel: jest.fn(),
            getReader: jest.fn(),
            pipeThrough: jest.fn(),
            pipeTo: jest.fn(),
            tee: jest.fn(),
          },
          blob: jest.fn(),
          arrayBuffer: jest.fn(),
          json: jest.fn(),
          text: jest.fn(),
          formData: jest.fn(),
          clone: jest.fn(),
          type: 'basic',
        },
      },
      mock,
    ]);

    return mock;
  }

  static findMockResponse(url: string, options: Readonly<RequestOptions>) {
    return Client.mockResponses.find(([response]) => {
      if (url !== response.url) {
        return false;
      }
      if ((options.method || 'GET') !== response.method) {
        return false;
      }
      return response.match.every(matcher => matcher(url, options));
    });
  }

  activeRequests: Record<string, typeof Request> = {};
  baseUrl = '';

  uniqueId() {
    return '123';
  }

  /**
   * In the real client, this clears in-flight responses. It's NOT
   * clearMockResponses. You probably don't want to call this from a test.
   */
  clear() {}

  wrapCallback<T extends any[]>(
    _id: string,
    func: FunctionCallback<T> | undefined,
    _cleanup: boolean = false
  ) {
    return (...args: T) => {
      if (hasProjectBeenRenamed(args)) {
        return;
      }
      respond(Client.mockAsync, func, ...args);
    };
  }

  requestPromise(
    path: string,
    {
      includeAllArgs,
      ...options
    }: {includeAllArgs?: boolean} & Readonly<RequestOptions> = {}
  ): any {
    return new Promise((resolve, reject) => {
      this.request(path, {
        ...options,
        success: (data, ...args) => {
          includeAllArgs ? resolve([data, ...args]) : resolve(data);
        },
        error: (error, ..._args) => {
          reject(error);
        },
      });
    });
  }

  static errors: Record<string, Error> = {};

  // XXX(ts): We type the return type for requestPromise and request as `any`. Typically these woul
  request(url: string, options: Readonly<RequestOptions> = {}): any {
    const [response, mock] = Client.findMockResponse(url, options) || [
      undefined,
      undefined,
    ];
    if (!response || !mock) {
      const methodAndUrl = `${options.method || 'GET'} ${url}`;
      // Endpoints need to be mocked
      const err = new Error(`No mocked response found for request: ${methodAndUrl}`);

      // Mutate stack to drop frames since test file so that we know where in the test
      // this needs to be mocked
      const lines = err.stack?.split('\n');
      const startIndex = lines?.findIndex(line => line.includes('.spec.'));
      err.stack = ['\n', lines?.[0], ...(lines?.slice(startIndex) ?? [])].join('\n');

      // Throwing an error here does not do what we want it to do....
      // Because we are mocking an API client, we generally catch errors to show
      // user-friendly error messages, this means in tests this error gets gobbled
      // up and developer frustration ensues.
      // We track the errors on a static member and warn afterEach test.
      Client.errors[methodAndUrl] = err;
    } else {
      // has mocked response

      // mock gets returned when we add a mock response, will represent calls to api.request
      mock(url, options);

      const body =
        typeof response.body === 'function' ? response.body(url, options) : response.body;

      if (![200, 202].includes(response.statusCode)) {
        response.callCount++;

        const errorResponse = Object.assign(
          {
            status: response.statusCode,
            responseText: JSON.stringify(body),
            responseJSON: body,
          },
          {
            overrideMimeType: () => {},
            abort: () => {},
            then: () => {},
            error: () => {},
          },
          new XMLHttpRequest()
        );

        this.handleRequestError(
          {
            id: '1234',
            path: url,
            requestOptions: options,
          },
          errorResponse as any,
          'error',
          'error'
        );
      } else {
        response.callCount++;
        respond(
          Client.mockAsync,
          options.success,
          body,
          {},
          {
            getResponseHeader: (key: string) => response.headers[key],
            statusCode: response.statusCode,
            status: response.statusCode,
          }
        );
      }
    }

    respond(Client.mockAsync, options.complete);
  }

  handleRequestError(
    {id, path, requestOptions}: HandleRequestErrorOptions,
    response: ResponseMeta,
    textStatus: string,
    errorThrown: string
  ) {
    const code = response?.responseJSON?.detail?.code;
    const isSudoRequired = code === SUDO_REQUIRED || code === SUPERUSER_REQUIRED;

    let didSuccessfullyRetry = false;

    if (isSudoRequired) {
      openSudo({
        isSuperuser: code === SUPERUSER_REQUIRED,
        sudo: code === SUDO_REQUIRED,
        retryRequest: async () => {
          try {
            const data = await this.requestPromise(path, requestOptions);
            requestOptions.success?.(data);
            didSuccessfullyRetry = true;
          } catch (err) {
            requestOptions.error?.(err);
          }
        },
        onClose: () =>
          // If modal was closed, then forward the original response
          !didSuccessfullyRetry && requestOptions.error?.(response),
      });
      return;
    }

    const errorCb = this.wrapCallback<[ResponseMeta, string, string]>(
      id,
      requestOptions.error
    );
    errorCb?.(response, textStatus, errorThrown);
  }
}

async function openSudo({onClose, ...args}: any = {}) {
  const mod = await import('sentry/components/modals/sudoModal');
  const {default: Modal} = mod;

  openModal(deps => <Modal {...deps} {...args} />, {onClose});
}

function openModal(renderer: (renderProps: any) => React.ReactNode, options?: any) {
  ModalStore.openModal(renderer, options ?? {});
}

export {Client};
