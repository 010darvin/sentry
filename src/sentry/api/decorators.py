from functools import wraps

from rest_framework.request import Request
from rest_framework.response import Response

from sentry.api.exceptions import EmailVerificationRequired, SudoRequired
from sentry.models import ApiKey
from sentry.models.apitoken import is_api_token_auth
from sentry.models.user import User
from sentry.services.hybrid_cloud.user import APIUser


def is_considered_sudo(request):
    # Right now, only password reauthentication (django-sudo) is supported,
    # so if a user doesn't have a password (for example, only has github auth)
    # then we shouldn't prompt them for the password they don't have.
    return (
        request.is_sudo()
        or isinstance(request.auth, ApiKey)
        or is_api_token_auth(request.auth)
        or request.user.is_authenticated
        and not request.user.has_usable_password()
    )


def sudo_required(func):
    @wraps(func)
    def wrapped(self, request: Request, *args, **kwargs) -> Response:
        # If we are already authenticated through an API key we do not
        # care about the sudo flag.
        if not is_considered_sudo(request):
            # TODO(dcramer): support some kind of auth flow to allow this
            # externally
            raise SudoRequired(request.user)
        return func(self, request, *args, **kwargs)

    return wrapped


def email_verification_required(func):
    @wraps(func)
    def wrapped(self, request: Request, *args, **kwargs) -> Response:
        if isinstance(request.user, User):
            if not request.user.get_verified_emails().exists():
                raise EmailVerificationRequired(request.user)
        elif isinstance(request.user, APIUser):
            has_verified_email = False
            for email in request.user.useremails:
                if email.is_verified:
                    has_verified_email = True
                    break
            if not has_verified_email:
                raise EmailVerificationRequired(request.user)
        else:
            raise NotImplementedError(
                "email_verification_required doesn't handle this type of user input"
            )
        return func(self, request, *args, **kwargs)

    return wrapped
