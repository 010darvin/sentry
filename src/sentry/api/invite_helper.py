from __future__ import annotations

from dataclasses import dataclass
from logging import Logger
from typing import Any, Dict

from django.utils.crypto import constant_time_compare
from rest_framework.request import Request

from sentry import features
from sentry.models import (
    Authenticator,
    AuthIdentity,
    AuthProvider,
    Organization,
    OrganizationMember,
    User,
    UserEmail,
)
from sentry.services.hybrid_cloud.user import APIUser
from sentry.signals import member_joined
from sentry.utils import metrics


def add_invite_details_to_session(request: Request, member_id: int, token: str) -> None:
    """Add member ID and token to the request session"""
    request.session["invite_token"] = token
    request.session["invite_member_id"] = member_id


def remove_invite_details_from_session(request: Request) -> None:
    """Deletes invite details from the request session"""
    request.session.pop("invite_member_id", None)
    request.session.pop("invite_token", None)


@dataclass(frozen=True, eq=True)
class InviteDetail:
    token: str | None
    member_id: int | None
    user_id: int | None
    is_authenticated: bool
    is_verified: bool


def get_invite_details(request: Request) -> InviteDetail:
    """Extracts invite details from request session"""
    user = request.user
    primary_email = UserEmail.objects.get_primary_email(user) if isinstance(user, User) else None

    return InviteDetail(
        token=request.session.get("invite_token", None),
        member_id=request.session.get("invite_member_id", None),
        user_id=user.id,
        is_authenticated=user.is_authenticated,
        is_verified=primary_email.is_verified if primary_email else False,
    )


class ApiInviteHelper:
    @classmethod
    def from_session_or_email(
        cls,
        detail: InviteDetail,
        organization: Organization,
        email: str,
        instance: Any | None = None,
        logger: Logger | None = None,
    ) -> ApiInviteHelper | None:
        """
        Initializes the ApiInviteHelper by locating the pending organization
        member via the currently set pending invite details in the session, or
        via the passed email if no cookie is currently set.
        """
        try:
            if detail.token and detail.member_id:
                om = OrganizationMember.objects.get(token=detail.token, id=detail.member_id)
            else:
                om = OrganizationMember.objects.get(
                    email=email, organization=organization, user=None
                )
        except OrganizationMember.DoesNotExist:
            # Unable to locate the pending organization member. Cannot setup
            # the invite helper.
            return None

        return cls(detail=detail, member_id=om.id, token=om.token, instance=instance, logger=logger)

    @classmethod
    def from_session(
        cls,
        request: Request,
        instance: Any | None = None,
        logger: Logger | None = None,
    ) -> ApiInviteHelper | None:
        detail = get_invite_details(request)

        if not detail.token or not detail.member_id:
            return None

        try:
            return ApiInviteHelper(
                detail=detail,
                member_id=detail.member_id,
                token=detail.token,
                instance=instance,
                logger=logger,
            )
        except OrganizationMember.DoesNotExist:
            if logger:
                logger.error("Invalid pending invite cookie", exc_info=True)
            return None

    def __init__(
        self,
        detail: InviteDetail,
        member_id: int,
        token: str | None,
        instance: Any | None = None,
        logger: Logger | None = None,
    ) -> None:
        if detail.member_id is not None and detail.member_id != member_id:
            raise ValueError("Mismatched member_id")
        if detail.token is not None and detail.token != token:
            raise ValueError("Mismatched token")

        self.detail = detail
        self.member_id = member_id
        self.token = token
        self.instance = instance
        self.logger = logger
        self.om = self.organization_member
        self.organization = self.om.organization

    def handle_success(self) -> None:
        member_joined.send_robust(
            member=self.om,
            organization=self.organization,
            sender=self.instance if self.instance else self,
        )

    def handle_member_already_exists(self) -> None:
        if self.logger:
            self.logger.info(
                "Pending org invite not accepted - User already org member",
                extra={"organization_id": self.organization.id, "user_id": self.detail.user_id},
            )

    def handle_member_has_no_sso(self) -> None:
        if self.logger:
            self.logger.info(
                "Pending org invite not accepted - User did not have SSO",
                extra={"organization_id": self.organization.id, "user_id": self.detail.user_id},
            )

    def handle_invite_not_approved(self) -> None:
        if not self.invite_approved:
            self.om.delete()

    @property
    def organization_member(self) -> OrganizationMember:
        return OrganizationMember.objects.select_related("organization").get(pk=self.member_id)

    @property
    def member_pending(self) -> bool:
        return self.om.is_pending  # type: ignore[no-any-return]

    @property
    def invite_approved(self) -> bool:
        return self.om.invite_approved  # type: ignore[no-any-return]

    @property
    def valid_token(self) -> bool:
        if self.token is None:
            return False
        if self.om.token_expired:
            return False
        tokens_are_equal = constant_time_compare(self.om.token or self.om.legacy_token, self.token)
        return tokens_are_equal  # type: ignore[no-any-return]

    @property
    def user_authenticated(self) -> bool:
        return self.detail.is_authenticated  # type: ignore[no-any-return]

    @property
    def member_already_exists(self) -> bool:
        if not self.user_authenticated:
            return False

        query = OrganizationMember.objects.filter(
            organization=self.organization, user_id=self.detail.user_id
        )
        return query.exists()  # type: ignore[no-any-return]

    @property
    def valid_request(self) -> bool:
        return (
            self.member_pending
            and self.invite_approved
            and self.valid_token
            and self.user_authenticated
            and not any(self.get_onboarding_steps().values())
        )

    def accept_invite(self, user: APIUser | None = None) -> OrganizationMember | None:
        om = self.om
        user_id = user.id if user else self.detail.user_id

        if self.member_already_exists:
            self.handle_member_already_exists()
            om.delete()
            return None

        try:
            provider = AuthProvider.objects.get(organization=om.organization)
        except AuthProvider.DoesNotExist:
            provider = None

        # If SSO is required, check for valid AuthIdentity
        if provider and not provider.flags.allow_unlinked:
            # AuthIdentity has a unique constraint on provider and user
            if not AuthIdentity.objects.filter(auth_provider=provider, user_id=user_id).exists():
                self.handle_member_has_no_sso()
                return None

        om.set_user_by_id(user_id)
        om.save()

        # TODO: Fix this
        # create_audit_entry(
        #     self.request,
        #     actor=user,
        #     organization=om.organization,
        #     target_object=om.id,
        #     target_user=user,
        #     event=audit_log.get_event_id("MEMBER_ACCEPT"),
        #     data=om.get_audit_log_data(),
        # )

        self.handle_success()
        metrics.incr("organization.invite-accepted", sample_rate=1.0)

        return om

    def _needs_2fa(self) -> bool:
        org_requires_2fa = self.organization.flags.require_2fa.is_set
        user_has_2fa = Authenticator.objects.user_has_2fa(self.detail.user_id)
        return org_requires_2fa and not user_has_2fa

    def _needs_email_verification(self) -> bool:
        return (
            features.has("organizations:required-email-verification", self.organization)
            and self.organization.flags.require_email_verification
            and not self.detail.is_verified
        )

    def get_onboarding_steps(self) -> Dict[str, bool]:
        return {
            "needs2fa": self._needs_2fa(),
            "needsEmailVerification": self._needs_email_verification(),
        }
