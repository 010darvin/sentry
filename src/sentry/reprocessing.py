from __future__ import absolute_import

from sentry.models import EventError


def report_processing_issue(event_data, scope, object=None, type=None, data=None):
    """Reports a processing issue for a given scope and object.  Per
    scope/object combination only one issue can be recorded where the last
    one reported wins.
    """
    if object is None:
        object = '*'
    if type is None:
        type = EventError.INVALID_DATA
    uid = '%s:%s' % (scope, object)
    event_data.setdefault('processing_issues', {})[uid] = {
        'scope': scope,
        'object': object,
        'type': type,
        'data': data,
    }
