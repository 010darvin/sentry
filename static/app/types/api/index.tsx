import {paths} from 'sentry/openapi';

export type ExtractPathResponseData<
  TPath extends keyof paths,
  TOperation extends keyof paths[TPath]
> = paths[TPath][TOperation] extends {
  responses: {200: {content: {'application/json': any}}};
}
  ? paths[TPath][TOperation]['responses'][200]['content']['application/json']
  : never;
