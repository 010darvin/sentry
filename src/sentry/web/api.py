from django.http import HttpResponse
from django.views.decorators.cache import cache_control
from django.views.generic.base import View as BaseView
from rest_framework.request import Request
from rest_framework.response import Response

from sentry.utils import json
from sentry.web.client_config import get_client_config


class ClientConfigView(BaseView):
    def get(self, request: Request) -> Response:
        return HttpResponse(json.dumps(get_client_config(request)), content_type="application/json")


@cache_control(max_age=3600, public=True)
def robots_txt(request):
    return HttpResponse("User-agent: *\nDisallow: /\n", content_type="text/plain")
