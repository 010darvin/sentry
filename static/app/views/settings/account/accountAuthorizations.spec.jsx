import {routerContext} from 'fixtures/js-stubs/routerContext';

import {render} from 'sentry-test/reactTestingLibrary';

import {Client} from 'sentry/api';
import AccountAuthorizations from 'sentry/views/settings/account/accountAuthorizations';

describe('AccountAuthorizations', function () {
  beforeEach(function () {
    Client.clearMockResponses();
  });

  it('renders empty', function () {
    Client.addMockResponse({
      url: '/api-authorizations/',
      method: 'GET',
      body: [],
    });

    const wrapper = render(<AccountAuthorizations />, {
      context: routerContext(),
    });

    expect(wrapper.container).toSnapshot();
  });
});
