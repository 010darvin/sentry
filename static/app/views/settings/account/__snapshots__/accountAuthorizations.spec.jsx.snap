// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountAuthorizations renders empty 1`] = `
<div>
  <div>
    <div
      class="css-q8xgg6-SettingsPageHeader e1nr98yc2"
    >
      <div
        class="css-1mfa9vw-TitleAndActions e1nr98yc8"
      >
        <div
          class="css-u3au9h-TitleWrapper e1nr98yc7"
        >
          <div
            class="css-18xqiae-Title e1nr98yc6"
          >
            <h4
              class="css-jz1c1u-HeaderTitle e12cn1y20"
            >
              Authorized Applications
            </h4>
          </div>
        </div>
      </div>
    </div>
    <p
      class="css-i23gaz-Description etd9r7m5"
    >
      <span>
        <span>
          You can manage your own applications via the 
        </span>
        <a
          class="css-ta9xa3-Link e1ycxor00"
          href="/settings/account/api/"
        >
          <span>
            API dashboard
          </span>
        </a>
        <span>
          .
        </span>
      </span>
    </p>
    <div
      class="css-2fj46w-Panel edxrl2k0"
    >
      <div
        class="css-1v7t0a0-PanelHeader-getPadding eknsyxy0"
      >
        Approved Applications
      </div>
      <div
        class="css-pemm1r-PanelBody-textStyles eszvord0"
      >
        <div
          class="css-fvblbb-EmptyMessage-EmptyMessage ege7ww24"
          data-test-id="empty-message"
        >
          <div
            class="css-u2uxsp-TextBlock-Description ege7ww21"
          >
            You haven't approved any third party applications.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
